Add parallel debugging section

dongahn opened the pull request:
- Add doc on how to debug parallel jobs using TotalView
Introduce basic TotalView commands
for launch and attach mode

- Add ways to improve Flux's own internal exec
events within TotalView.

- Add known issues section to document
current limitations.

- Increase flux-core version number to 0.19.0

SteVwonder commented on conf.py:
```diff
@@ -34,7 +34,7 @@
 
 # The full version, including alpha/beta/rc tags
-release = '0.13.0'
+release = '0.19.0'
```
We should probably delete this line.  We did so in flux-core, and I don't think this is used anywhere.  @gonsie, can you confirm/do you agree?

If we don't delete it, we should set it to `0.1.0` or `0.0.0` since flux-docs will be versioned independently from flux-core and flux-sched.

dongahn replied:
Yes, too much churn if we have to update this each time. The version number shows up in the browser window and it was confusing. 

Do we need a way to sync the html doc version with flux-core version though?

dongahn replied:
@SteVwonder and @gonsie: just recommend what you need with the version number, I will adjust the PR.

SteVwonder replied:
> Do we need a way to sync the html doc version with flux-core version though?

From previous conversations in coffee time and issues, I think the conclusion we came to is:
"Each repo has it’s own documentation and that gets tagged and released along with the code, but the high-level “meta” documentation has it’s own versioning that is divorced from any particular sub-packages/repos versioning."

So the top-level docs can refer to either the latest version of flux-core docs or a specific tag (if necessary) via Sphinx links.  But the versioning of flux-docs itself is entirely independent.  We still need to come up with conventions for how and when we want to tag flux-docs.

SteVwonder replied:
> @SteVwonder and @gonsie: just recommend what you need with the version number, I will adjust the PR.

What happens if we just wholesale delete the line of code like was done in flux-core? I think it should just not version anything, and we will just get `latest`.  Which seems fine to me for now until we come up with a convention for tagging and releasing flux-docs.  Thoughts?

dongahn replied:
> So the top-level docs can refer to either the latest version of flux-core docs or a specific tag (if necessary) via Sphinx links. But the versioning of flux-docs itself is entirely independent. We still need to come up with conventions for how and when we want to tag flux-docs.

Very wise decision! Thanks.

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,84 @@
+Debugging Jobs 
+==============
+
+Flux supports parallel debuggers such as Rogue Wave Software (RWS)'s
+TotalView parallel debugger (https://www.roguewave.com).
```
ReST has support for external links.  The syntax is weird:

```suggestion
Flux supports parallel debuggers such as `Rogue Wave Software (RWS)'s <https://www.roguewave.com>`_
TotalView parallel debugger.
```

dongahn replied:
Learned a new thing! Will change.

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,84 @@
+to attach to your running code if you just attach it
+to the ``flux mini run`` process that controls your code.
+
+Flux provides ``MPIR_partial_attach_ok`` (https://www.mpi-forum.org/docs/mpir-specification-10-11-2010.pdf)
```
I **think** it is possible to combine the verbatim mode (``) with hyperlinks, but I don't want to suggest something without trying it first.  If it isn't easily doable, maybe something like:

```suggestion
Flux provides ``MPIR_partial_attach_ok`` `[1]<https://www.mpi-forum.org/docs/mpir-specification-10-11-2010.pdf>`_
```

dongahn replied:
OK. Will try that.

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,84 @@
+Flux provides ``MPIR_partial_attach_ok`` (https://www.mpi-forum.org/docs/mpir-specification-10-11-2010.pdf)
+so you can attach TotalView to a subset of parallel processes as well, which
+can be handy when you debug a large-scale job. Please refer to
+TotalView user guide on its subset-attach capability: https://docs.roguewave.com/en/totalview/current/html/.
```
Since the TotalView user guide is referenced multiple times, you can created a named link rather than an anonymous one (I **think** this name is valid anywhere in this file, maybe even across the entire repo).

```suggestion
`TotalView user guide`_ on its subset-attach capability.

.. _TotalView user guide: https://docs.roguewave.com/en/totalview/current/html/
```

dongahn replied:
Good idea. Just to make sure: here, the external hyperlink is given with the "normal" syntax?

SteVwonder replied:
The Sphinx documentation describes it better than I can: https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#external-links

There is basically two ways to do it: with an inline/anonymous link or with a "ref" and "target definition". For this particular case, I recommend trying the latter since this link it used multiple times. You can define the target once, and then reference it multiple times.

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,84 @@
+to enable this only for the user's own sessions.
+Notice that it is designed to support not only Flux but also SLURM's
+srun and IBM JSM's jsrun commands. The ``regex`` syntax of
+``exec_handling`` within TotalView can be found in the TotalView reference
+guide: e.g., https://docs.roguewave.com/en/totalview/current/html/.
```
Assuming the named reference suggested above is created:

```suggestion
``exec_handling`` within TotalView can be found in the `TotalView user
guide`_.
```

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,84 @@
+Debugging your MPI job from the beginning of
+parallel program execution:
+
+.. code-block:: sh
+
+  totalview --args flux mini run -N 2 -n 2 ./mpi-program
```
Nit pick.  ReST has a `console` code-block that supports highlighting of the terminal prompt, command, and output.  Since this is just a single line, it doesn't matter all that much.  Mainly an FYI.

```suggestion
.. code-block:: console

  $ totalview --args flux mini run -N 2 -n 2 ./mpi-program
```

dongahn replied:
Nice! Pretty powerful. Does ReST support most of GitHub mark down verbatim block ? In this case, for example ```console?

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,84 @@
+
+  totalview --args flux mini attach <JOBID>
+
+Note that you should use ``totalview --args flux mini attach`` whether your
+job has been launched via ``flux mini run`` or ``flux mini submit``.
+In other word, Flux currently does not allow TotalView
```
```suggestion
Note that you should use ``totalview --args flux mini attach`` when your
job has been launched via ``flux mini run`` or ``flux mini submit``.
In other words, Flux currently does not allow TotalView
```

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,84 @@
+
+  totalview --args flux mini attach <JOBID>
+
+Note that you should use ``totalview --args flux mini attach`` whether your
```
ReSt has a note directive, if you want to use that:

```suggestion
.. note::
You should use ``totalview --args flux mini attach`` whether your
```

dongahn replied:
Cool!

dongahn commented:
@petertea: Moving the discussion to GitHub PR. 

> A quick comment.  The links for TotalView still point to the RogueWave site,
which do still work, but I wonder if you might want to update those to reflect
Perforce and the new web site,

> https://totalview.io

Yes I will do this. 

> The doc links still works since we haven't moved those (yet).

> I'm trying a test on butte, where a new version of Spectrum MPI has been
installed, and I'm getting a Unable to connect to Flux, and it's an errno 2,
ENOENT, so it seems like it's not finding o   


> This is just trying to run flux.
I haven't brought TotalView into the mix yet.   I'll take a closer look at the
complete documentation to see if I'm missing something obvious.

You might find the following doc useful: https://flux-framework.readthedocs.io/en/latest/coral.html


> Overall the addition makes sense to me as how to run TotalView. The
exec_handling string might need a bit of work to make it more compact, but I'd
need to play around with it a bit before I can make a recommendation.

Cool.

dongahn commented:
@SteVwonder or @gonsie: This PR should be ready. I addressed all of your comments and incorporated other findings that came out from testing and talking to TotalView folks. This is based on some changes in [this PR](https://github.com/flux-framework/flux-core/pull/3130). So it would be wise to land that PR first.

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,101 @@
+it unless you are a Flux developer!
+Instead, you can make TotalView automatically
+continue on these events by building the the following
+exec-handling Tcl code into TotalView.
+
+.. code-block::
+
+    catch {dset TV::exec_handling { {^(flux|lrun|srun|jsrun)(<python[^>]*>(<flux-job>)?|<bash>|<jsrun>|<jswrap>)*$ go}} }
```
Unfortunately, this block is not rendering in readthedocs right now.  I wonder if it is because the language is missing (although the docs say in Sphinx 2.0 it became optional):

```suggestion
exec-handling Tcl code into TotalView.

.. code-block:: tcl

    catch {dset TV::exec_handling { {^(flux|lrun|srun|jsrun)(<python[^>]*>(<flux-job>)?|<bash>|<jsrun>|<jswrap>)*$ go}} }
```

If that doesn't work, it looks like you can get a literal block without the code block directive by ending the previous line with `::` and indenting the actual block [[1]](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks):

```suggestion
exec-handling Tcl code into TotalView::

    catch {dset TV::exec_handling { {^(flux|lrun|srun|jsrun)(<python[^>]*>(<flux-job>)?|<bash>|<jsrun>|<jswrap>)*$ go}} }
```

dongahn replied:
I tried with Tcl and it didn't work. I will just go with the second suggestion.

SteVwonder commented on debugging.rst:
```diff
@@ -0,0 +1,101 @@
+==============
+
+Flux supports parallel debuggers such as Rogue Wave Software (RWS)'s
+TotalView parallel debugger (https://totalview.io).
```
Nitpick. Feel free to ignore:

```suggestion
`TotalView parallel debugger <(https://totalview.io)>`_.
```

dongahn commented:
OK. Made those suggested changes. I just also put MWP label.
